wdl-lsp/tests/common.rs

```diff
@@ -0,0 +1,292 @@
+}
+
+const MAX_BUF_SIZE: usize = 4096;
+
+impl TestContext {
+    /// Crates a new test context.
+    ///
+    /// The `base` parameter is the name of a subdirectory in `tests/workspace`
+    /// which contains the WDL files for the test. These files are copied
+    /// into a temporary workspace directory.
+    pub fn new(base: &str) -> Self {
+        let (request_tx, req_server) = duplex(MAX_BUF_SIZE);
+        let (resp_server, response_rx) = duplex(MAX_BUF_SIZE);
+        let response_rx = BufReader::new(response_rx);
+
+        let (service, socket) = LspService::new(|client| {
+            Server::new(
+                client,
+                ServerOptions {
+                    lint: true,
+                    ..Default::default()
+                },
+            )
+        });
+        let server =
+            tokio::spawn(tower_lsp::Server::new(req_server, resp_server, socket).serve(service));
+
+        let workspace = TempDir::new().unwrap();
+        let workspace_path = get_workspace_path(base);
+        if workspace_path.exists() {
+            let items: Vec<_> = fs::read_dir(&workspace_path)
+                .unwrap()
+                .map(|e| e.unwrap().path())
+                .collect();
+            let options = CopyOptions::new().overwrite(true);
+            copy_items(&items, workspace.path(), &options).unwrap();
+        }
+
+        Self {
+            request_tx,
+            response_rx,
+            _server: server,
+            request_id: 0,
+            workspace,
+        }
+    }
+
+    /// Creates a file URI for a path within the temporary workspace.
+    pub fn doc_uri(&self, path: &str) -> Url {
+        Url::from_file_path(self.workspace.path().join(path)).unwrap()
+    }
+
+    /// Sends a raw JSON-RPC request to the server.
+    pub async fn send_raw(&mut self, message: &str) {
+        self.request_tx
+            .write_all(encode_message(message).as_bytes())
+            .await
+            .unwrap();
+    }
+
+    /// Sends a typed JSON-RPC request to the server.
+    pub async fn send(&mut self, request: &jsonrpc::Request) {
+        let content = serde_json::to_string(request).unwrap();
+        self.send_raw(&content).await;
+    }
+
+    /// Reads a raw JSON-RPC message string from the server response stream.
+    pub async fn read_message_str(&mut self) -> Option<String> {
+        let mut content_length = 0;
+
+        loop {
+            let mut header = String::new();
+            if self.response_rx.read_line(&mut header).await.unwrap() == 0 {
+                return None; // Connection closed
+            }
+            if header.trim().is_empty() {
+                break; // End of headers
+            }
+            let parts: Vec<&str> = header.trim().splitn(2, ": ").collect();
+            if parts.len() == 2 && parts[0].eq_ignore_ascii_case("Content-Length") {
+                content_length = parts[1].parse().unwrap();
+            }
+        }
+
+        if content_length > 0 {
```

Comment:
Technically speaking, I think a `content_length` of `0` would return an empty `String`. Maybe we should update that.